### DISCUSSION TITLE
Realtek Ameba Micropython (BW16)

### DISCUSSION THREAD

**Discussion by mvdw73** [micropython/micropython#11979] [Other]
Thanks to the great work by Matty Trentini and others, there is a micropython port for the Realtek Ameba series of chips. I have been playing around with the AI Thinker BW16 dev board with the RTL8720DN on board, and am having some issues with trying to get a basic MQTT pub/sub client going.

I have uploaded the umqtt.simple code to my device, but I have run into the following issues:

* The ameba port has a non-standard socket interface. It uses recv/send instead of read/write. Luckily I was able to fix that fairly easily.
* When I try to run the connect() method of the MQTTClient class, I'm failing on the poll.register() function, with the following error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: stream operation not supported
```

I can reproduce this error in a standalone program simply as follows:

```
>>> import usocket as socket
>>> import uselect
>>> s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>>> s.connect("radarbroker.wlan", 1883)

Connect to Server successfully!
client connect successfully
True
>>> p = uselect.poll()
>>> p.register(s, uselect.POLLIN)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: stream operation not supported
>>> 
```

I'm not sure whether this is related to the non-standard socket interface in the port, or whether there is a further issue in the uselect module compiled into the ameba port.

The github source I'm referring to is the ameba branch of the repo at https://github.com/ambiot/micropython (which somewhat concerningly is 1586 commits behind micrpython:master)

This issue happens whether I am running the "release" version (V1.1.0 in the repo) of the micropython firmware, or a build from github updated today.

**Comment by bidrohini** [comment]
You can try umqtt.robust or paho.mqtt instead.

**Reply by mvdw73** [reply]
paho mqtt is probably no good for me - I'm reluctant to even try that on my board given the non standard socket interface.

Probably a better place to start will be to get usocket on the ameba port the same as the others or at least the same as the ESP32 port.

Then I can look into the poll() interface, and see if that's where it's falling down.

[EDIT]: after consultation with Matt Trentini, I have decided to concentrate my efforts on the usocket port for the Ameba chips. I will start with looking at the RP2 port rather than the ESP32 - advice is that the RP2 port is more up to date, and was written by Damien George so should be well integrated with the main codebase.
